mountpoint-s3-client/src/mock_client/throughput_client.rs

```diff
@@ -57,23 +59,41 @@ impl ThroughputMockClient {
 }
 
 #[pin_project]
-pub struct GetObjectResult {
+pub struct ThroughputGetObjectRequest {
     #[pin]
-    inner: BoxStream<'static, ObjectClientResult<GetBodyPart, GetObjectError, MockClientError>>,
+    request: MockGetObjectRequest,
+    rate_limiter: LeakyBucket,
 }
 
-impl Stream for GetObjectResult {
+impl GetObjectRequest for ThroughputGetObjectRequest {
+    type ClientError = MockClientError;
+
+    fn increment_read_window(self: Pin<&mut Self>, len: usize) {
+        let this = self.project();
+        this.request.increment_read_window(len);
+    }
+}
+
+impl Stream for ThroughputGetObjectRequest {
     type Item = ObjectClientResult<GetBodyPart, GetObjectError, MockClientError>;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let this = self.project();
-        this.inner.poll_next(cx)
+        this.request.poll_next(cx).map(|next| {
+            next.map(|item| {
+                item.map(|body_part| {
+                    // Acquire enough tokens for the number of bytes we want to deliver
+                    block_on(this.rate_limiter.acquire(body_part.1.len() as u32));
```

Comment:
This looks a little bit more low-level then it is was before, is it required? Previously we were awaiting `acquire` in a stream wrapped by `then`, now we `async_io::block_on` it. It looks equivalent, but a little trickier to follow